### ISSUE TITLE
[docs] Incorrect pseudocode for automatic optimization with multiple optimizers

### ISSUE THREAD

**Issue by nik-sm** [PyTorchLightning/pytorch-lightning#8525]
I have a model with two optimizers. 

The docs at https://pytorch-lightning.readthedocs.io/en/latest/common/optimizers.html#automatic-optimization currently state that when training with multiple models and automatic optimization, each batch will be run through BOTH models.

However, by using using `Trainer(..., limit_train_batches=10)` and counting the number of batches seen by each model, I observe that:
- with a single model present, it will see 10 batches (making 10 gradient steps)
- with two models present, each will see 5 batches (making 5 gradient steps each).

I think it would be useful and preferred for the training logic to follow the pseudocode currently in the docs. 
Consider in my case, for example. The ideal "control experiment" to ensure I have incorporated my secondary model correctly is to vary the weight on its loss term from 0 upwards.
At a weight of 0, I should see identical behavior between the case of { secondary model not present } and the case of { secondary model present, with multiplier 0 }. 
Basiclaly, the only effect of having the secondary model in this control experiment would be some useless forward and backward passes that do not change any layer weights.

However, with the current behavior, I don't think there is any way to get such an identical control comparison.

If I try running my model for twice as many epochs to compensate, I will see "statistically similar" results, but not identical ones.
When the secondary model is not present, in 1 epoch the main model receives 10 pseudorandom batches from the dataset.
When the secondary model is present (and its multiplier is 0, the control scenario), in 2 epochs the main model receives 5 pseudorandom batches twice.
In other words, it sees 10 batches both times, but they are not the same batches.

In the meantime, it would be great if the docs more accurately captured the current behavior, for example:
```python
for epoch in epochs:
    for batch_idx, batch in enumerate(data):
        opt_idx = idx % len(optimizers)
        opt = optimizers[opt_idx]
        loss = model.training_step(batch, batch_idx, opt_idx)
        opt.zero_grad()
        loss.backward()
        opt.step()

    for lr_scheduler in lr_schedulers:
        lr_scheduler.step()
```

Happy to discuss further if my explanation above is not clear.

Cheers!

(p.s. I fixed the pseudocode below to be more accurate to my actual training logic; in the previous version I posted as a github discussion, the `loss2` term would have no effect on `self.main_model`)

### Discussed in https://github.com/PyTorchLightning/pytorch-lightning/discussions/8516

<div type='discussions-op-text'>

<sup>Originally posted by **nik-sm** July 21, 2021</sup>
Hi - I'm seeing unexpected behavior with automatic optimization.
I have two `nn.Modules` in my model, and in order to understand the effect of that second model, I'm trying some experiments with and without it.
The setup is something like this:

```python
def configure_optimizers(self):
    if self.secondary_model:
        return [Adam(self.main_model.parameters()), Adam(self.secondary_model.parameters()]
    else:
        return Adam(self.main_model.parameters())

def training_step(self, batch, batch_idx, optimizer_idx=0):
    data, labels = batch
    preds = self.main_model(data)
    loss1 = f(preds, labels)
    if self.secondary_model:
        loss2 = g(self.secondary_model(preds), labels)
    if self.secondary_model and optimizer_idx > 0:
        return alpha * loss2
    else:
        return loss1 + alpha * loss2
 ```

**Observation 1**: the code runs significantly faster with the second model included, which surprises me.
The secondary model is smaller, so if using two optimizers meant that 1/2 of gradient steps were only using this smaller model, then I could understand.
However the algorithm described in the docs at https://pytorch-lightning.readthedocs.io/en/latest/common/optimizers.html#automatic-optimization) should result in the computation being equal or greater (and never less):
```
In the case of multiple optimizers, Lightning does the following:

 for epoch in epochs:
     for batch in data:
         for opt in optimizers:
             loss = model.training_step(batch, batch_idx, optimizer_idx)
             opt.zero_grad()
             loss.backward()
             opt.step()
     for lr_scheduler in lr_schedulers:
         lr_scheduler.step()
```

**Observation 2**: if I set that coefficient `alpha=0` , the main model's performance is NOT identical.
Based on the pseudocode from the docs above, the main model should see all the same batches and perform exactly the same number of gradient steps.
I am using `deterministic=True` in the trainer object, and setting seed at beginning of my script with `seed_everything(1, workers=True)`.
There is no randomness in the secondary model, so random seed should not be affected by that inner `for opt in optimizers loop`.

Am I misunderstanding how automatic optimization works, or is there something else that might be affecting me here?

Thanks in advance!</div>

**Comment by nik-sm** [comment]
@carmocca

**Comment by carmocca** [comment]
> when training with multiple models and automatic optimization

Did you mean multiple optimizers?

> and counting the number of batches seen by each model, I observe that:

With this reproduction snippet, it does appear that each batch is used for all available optimizers

```python
import os

import torch
from torch.utils.data import DataLoader, Dataset

from pytorch_lightning import LightningModule, Trainer


class RandomDataset(Dataset):

    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return {'i': index, 'data': self.data[index]}

    def __len__(self):
        return self.len


class BoringModel(LightningModule):

    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx, optimizer_idx):
        batch, i = batch['data'], batch['i'].item()
        current_epoch = self.trainer.current_epoch
        print(f"{current_epoch=}, {batch_idx=}, {i=}, {optimizer_idx=}")
        loss = self(batch).sum()
        self.log("train_loss", loss)
        return {"loss": loss}

    def configure_optimizers(self):
        return [
            torch.optim.SGD(self.layer.parameters(), lr=0.1),
            torch.optim.SGD(self.layer.parameters(), lr=0.1)
        ]


def run():
    train_data = DataLoader(RandomDataset(32, 64), batch_size=1)

    model = BoringModel()
    trainer = Trainer(
        default_root_dir=os.getcwd(),
        limit_train_batches=3,
        max_epochs=3,
        weights_summary=None,
        progress_bar_refresh_rate=0,
    )
    trainer.fit(model, train_dataloaders=train_data)


if __name__ == '__main__':
    run()
```

output:
```python
current_epoch=0, batch_idx=0, i=0, optimizer_idx=0
current_epoch=0, batch_idx=0, i=0, optimizer_idx=1
current_epoch=0, batch_idx=1, i=1, optimizer_idx=0
current_epoch=0, batch_idx=1, i=1, optimizer_idx=1
current_epoch=0, batch_idx=2, i=2, optimizer_idx=0
current_epoch=0, batch_idx=2, i=2, optimizer_idx=1
current_epoch=1, batch_idx=0, i=0, optimizer_idx=0
current_epoch=1, batch_idx=0, i=0, optimizer_idx=1
current_epoch=1, batch_idx=1, i=1, optimizer_idx=0
current_epoch=1, batch_idx=1, i=1, optimizer_idx=1
current_epoch=1, batch_idx=2, i=2, optimizer_idx=0
current_epoch=1, batch_idx=2, i=2, optimizer_idx=1
current_epoch=2, batch_idx=0, i=0, optimizer_idx=0
current_epoch=2, batch_idx=0, i=0, optimizer_idx=1
current_epoch=2, batch_idx=1, i=1, optimizer_idx=0
current_epoch=2, batch_idx=1, i=1, optimizer_idx=1
current_epoch=2, batch_idx=2, i=2, optimizer_idx=0
current_epoch=2, batch_idx=2, i=2, optimizer_idx=1
```

Can you try to reproduce the different loop behaviour you are experiencing with this example?

**Comment by nik-sm** [comment]
Thanks for following up. This weekend I can I can try to shrink my code down to a minimal example and add here. 

Yes I did mean multiple optimizers, not just multiple models.

Just as a sanity check, I confirmed my observation by just doing the following: 

```python
def __init__(self, ...):
    ...
    self.batch_counts = [0, 0]

def training_step(..., optimizer_idx=0):
    ...
    self.batch_counts[optimizer_idx] += 1

trainer.fit(model, datamodule=...)
print(model.batch_counts)
```

And as you say, for 3 batches of data:
- your snippet above gives `batch_counts = [3, 3]` (matching the docs)
- my code gives `batch_counts = [2, 1]`.

Talk soon.

**Comment by nik-sm** [comment]
Sigh. After a bit of a saga, I finally tracked down the reason for this behavior. I was using:

```python
def configure_optimizers(self):
    result = [{"optimizer": main_optimizer, "frequency": 1}]
    if self.use_second_model:
        ...
        result.append({"optimizer": adv_optimizer, "frequency": 1})
```

This `frequency` keyword has some special behavior, and after a lot of googling I found some fine print in the docs at https://pytorch-lightning.readthedocs.io/en/1.3.8/common/lightning_module.html#configure-optimizers:

```
The frequency value specified in a dict along with the optimizer key is an int corresponding to the number of sequential batches optimized with the specific optimizer. 
It should be given to none or to all of the optimizers. 
**There is a difference between passing multiple optimizers in a list, and passing multiple optimizers in dictionaries with a frequency of 1**: 
In the former case, all optimizers will operate on the given batch in each optimization step. 
In the latter, only one optimizer will operate on the given batch at every step. 
This is different from the frequency value specified in the lr_dict mentioned below.
```

My natural assumption would have been: I'd like to experiment with different scheduling frequencies, like 1:1, 1:2, etc - so I'll start with 1:1 and then scale up as needed.

But based on the magic behavior of this keyword, in fact what I would need to do is something a bit more goofy like
```python
def configure_optimizers(self):
    result = []
    main_opt_dict = {"optimizer": main_optimizer}
    if self.use_second_model and self.second_opt_frequency is not None:
        main_opt_dict["frequency"] = 1
    result.append(main_opt_dict)
    if self.use_second_model:
        second_opt_dict = {"optimizer": adv_optimizer}
        if self.second_opt_frequency is not None:
             second_opt_dict["frequency"] = self.second_opt_frequency
        result.append(second_opt_dict)
    return result
```

Note also that on the page https://pytorch-lightning.readthedocs.io/en/latest/common/optimizers.html#automatic-optimization it makes a very clear and simple statement:

> Under the hood, Lightning does the following:

but in fact, this is not true - the pseudocode that follows is highly dependent on the return structure from `configure_optimizers`, and one must also carefully read https://pytorch-lightning.readthedocs.io/en/latest/common/lightning_module.html#configure-optimizers to achieve the specified behavior.

---

Let me begin by thanking you for the help and attention here.
I know that this issue may not be the ideal place to give higher-level feedback, but I figure I might as well since you've been very receptive and helpful with this particular issue, and it's feedback I would want to hear if I was developing the library.

I really really love the idea behind pytorch-lightning. 
There is too much time spent on boilerplate and too little reproducibility in research code!
But this experience (which in a strict technical sense was due to my own user error) is a little bit emblematic of my overall experience with the library: 
If everything goes well, the great features and ideas will save me time and effort - but if I miss just a few bits of the fine-print, things will go wrong in a way that is very difficult to debug, resulting in a net loss of development time!


To give just one other example - elsewhere in my current project, I lost a similar amount of time to an analogous situation with hyperparameter saving into checkpoints. 
You can see all throughout the examples here https://pytorch-lightning.readthedocs.io/en/stable/common/hyperparameters.html
that it seems the simple approach of passing an `args.Namespace` object into a model is a reasonable and recommended approach.
```python
# using a argparse.Namespace
class LitMNIST(LightningModule):
    def __init__(self, hparams, *args, **kwargs):
        super().__init__()
```
However, in https://pytorch-lightning.readthedocs.io/en/stable/common/weights_loading.html#checkpoint-loading you can see that this is actually NOT the right approach (or at least has very important caveats)
> If your model’s hparams argument is Namespace and .yaml file has hierarchical structure, you need to refactor your model to treat hparams as dict.

(Specifically, storing an `argparse.Namespace` object results in `self.args` and then doing `self.save_hyperparams()` will successfully save a `hparams.yaml` and will also successfully save these params into the `*.ckpt` file. But for some reason, these params will NOT be loaded automatically loaded from the checkpoint).

---

Anyway - thanks again for your help and attention with debugging this particular issue, and I hope there is some way to continue making the docs more consistent/prevent pitfalls like these.

**Comment by nik-sm** [comment]
A quick follow-up as I change my code and remove this `frequency` keyword for now.

When there are multiple optimizers used in `training_step`, then it seems that `training_epoch_end` receives a list with an extra layer of nesting, like:
```python
outputs = [
  [{...},  # batch1, opt1
   {...}], # batch1, opt2
  [{...},  # batch2, opt1
   {...}]  # batch2, opt2
...
]
```
The docs at https://pytorch-lightning.readthedocs.io/en/latest/common/lightning_module.html#training-epoch-end are incorrect and do not mention this

**Comment by carmocca** [comment]
> and it's feedback I would want to hear if I was developing the library.

Do not hesitate! Any feedback is very welcome!

I see all these cuts and inconsistencies you are suffering as opportunities for docs improvements. This way any future readers have a better chance of not losing the same time. So feel free to send PRs with improvements for all these!

> but in fact, this is not true - the pseudocode that follows is highly dependent on the return structure from configure_optimizers, and one must also carefully read https://pytorch-lightning.readthedocs.io/en/latest/common/lightning_module.html#configure-optimizers to achieve the specified behavior.

Here, we can mention that optimizers with frequencies defined can alter the follow and link to the `configure_optimizers` docs.

> However, in https://pytorch-lightning.readthedocs.io/en/stable/common/weights_loading.html#checkpoint-loading you can see that this is actually NOT the right approach (or at least has very important caveats)

It's not a wrong approach, but as you found out, we require that you access the `hparams` with the `__getitem__` operator (`[]` access). This is because when we `yaml.load()` the hparams file, we don't know which class generated this hparams file in particular so a `dict` is returned.

However, this would suggest that `argparse.Namespace` is subscriptable, which is not.
So this needs another docs update to suggest converting the `Namespace` into an `AttributeDict`.

I will open a PR with this.

> are incorrect and do not mention this.

Agree!

**Comment by nik-sm** [comment]
Thanks for the very positive attitude :)

> This is because when we yaml.load() the hparams file, we don't know which class generated this hparams file in particular so a dict is returned.

A quick thought - would using https://pypi.org/project/dotted-dict/ or a similar class for the returned container class be helpful?

It's basically just a subclass of dict that maps `__getattr__` to `__getitem__`, and `__setattr__` to `__setitem__`, etc, so that the contents can be access either with `container["key"]` or `container.key`.

**Comment by carmocca** [comment]
> A quick thought - would using https://pypi.org/project/dotted-dict/ or a similar class for the returned container class be helpful?

We already have this:
https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/utilities/parsing.py#L255

I opened a docs update in #8577 to show how to properly pass a `Namespace` to the `LightningModule` as a config.
